File: services/api/internal/auth/roles.go
package auth

import (
	"net/http"
	"slices"
	"strings"

	"github.com/therealagt/ContractManagementTool/services/api/internal/config"
)

type Role string

const (
	RoleUploader Role = "uploader"
	RoleReviewer Role = "reviewer"
	RoleAuditor  Role = "auditor"
	RoleAdmin    Role = "admin"
)

var allRoles = []Role{RoleUploader, RoleReviewer, RoleAuditor, RoleAdmin}

func RolesForUser(settings *config.Settings, user *IAPUser) []Role {
	email := strings.ToLower(user.Email)
	if _, ok := settings.AdminEmails()[email]; ok {
		return slices.Clone(allRoles)
	}

	var roles []Role
	if _, ok := settings.UploaderEmails()[email]; ok {
		roles = append(roles, RoleUploader)
	}
	if _, ok := settings.ReviewerEmails()[email]; ok {
		roles = append(roles, RoleReviewer)
	}
	if _, ok := settings.AuditorEmails()[email]; ok {
		roles = append(roles, RoleAuditor)
	}
	return roles
}

func HasAnyRole(granted []Role, required ...Role) bool {
	for _, role := range required {
		if slices.Contains(granted, role) {
			return true
		}
	}
	return false
}

func AssertSeparationOfDuty(actor, other, action string) error {
	if strings.EqualFold(actor, other) {
		return &SODViolation{Action: action}
	}
	return nil
}

type SODViolation struct {
	Action string
}

func (e *SODViolation) Error() string {
	return "separation of duties violation: same actor cannot " + e.Action
}

func WriteForbidden(w http.ResponseWriter, message string) {
	http.Error(w, message, http.StatusForbidden)
}

func WriteUnauthorized(w http.ResponseWriter, message string) {
	http.Error(w, message, http.StatusUnauthorized)
}

File: services/api/app/auth/roles.py
from enum import StrEnum

from fastapi import Depends, HTTPException, status

from app.auth.dependencies import get_iap_user
from app.auth.iap import IapUser
from app.config import settings


class Role(StrEnum):
    UPLOADER = "uploader"
    REVIEWER = "reviewer"
    AUDITOR = "auditor"
    ADMIN = "admin"


_ALL_ROLES = frozenset(Role)


def roles_for_user(user: IapUser) -> frozenset[Role]:
    email = user.email.lower()
    roles: set[Role] = set()

    if email in settings.admin_emails:
        return _ALL_ROLES
    if email in settings.uploader_emails:
        roles.add(Role.UPLOADER)
    if email in settings.reviewer_emails:
        roles.add(Role.REVIEWER)
    if email in settings.auditor_emails:
        roles.add(Role.AUDITOR)

    return frozenset(roles)


def require_roles(*required: Role):
    def dependency(user: IapUser = Depends(get_iap_user)) -> IapUser:
        granted = roles_for_user(user)
        if not any(role in granted for role in required):
            raise HTTPException(
                status.HTTP_403_FORBIDDEN,
                "Insufficient permissions for this action",
            )
        return user

    return dependency


def assert_separation_of_duty(*, actor: str, other: str, action: str) -> None:
    if actor.lower() == other.lower():
        raise HTTPException(
            status.HTTP_403_FORBIDDEN,
            f"Separation of duties violation: same actor cannot {action}",
        )
